plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/action/DefaultActionDispatcher.java

```diff
@@ -15,44 +15,198 @@
  ********************************************************************************/
 package org.eclipse.glsp.server.action;
 
+import java.util.Collections;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
 import java.util.stream.Collectors;
 
 import org.apache.log4j.Logger;
 import org.eclipse.glsp.api.action.Action;
 import org.eclipse.glsp.api.action.ActionDispatcher;
+import org.eclipse.glsp.api.action.ActionMessage;
 import org.eclipse.glsp.api.action.kind.ResponseAction;
 import org.eclipse.glsp.api.handler.ActionHandler;
 import org.eclipse.glsp.api.model.GraphicalModelState;
 import org.eclipse.glsp.api.model.ModelStateProvider;
+import org.eclipse.glsp.api.protocol.ClientSessionListener;
+import org.eclipse.glsp.api.protocol.ClientSessionManager;
+import org.eclipse.glsp.api.protocol.GLSPClient;
 import org.eclipse.glsp.api.registry.ActionHandlerRegistry;
 
 import com.google.inject.Inject;
+import com.google.inject.Provider;
 
-public class DefaultActionDispatcher implements ActionDispatcher {
-   private static Logger LOG = Logger.getLogger(DefaultActionDispatcher.class);
+/**
+ * <p>
+ * An ActionDispatcher that executes all handlers in the same thread. The dispatcher's
+ * public methods can be invoked from any thread, e.g. from background jobs running on
+ * the server.
+ * </p>
+ */
+public class DefaultActionDispatcher implements ActionDispatcher, ClientSessionListener {
 
+   private static final Logger LOG = Logger.getLogger(DefaultActionDispatcher.class);
+
+   private static final AtomicInteger COUNT = new AtomicInteger(0);
+
+   @Inject
+   private ActionHandlerRegistry actionHandlerRegistry;
```

Comment:
Please use protected instead of private for properties and methods that will potentially be used/customized by subclasses.